opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -1175,7 +1181,82 @@ namespace Opm {
         return {more_network_update, well_group_control_changed};
     }
 
+    //PJPE: This function is to be used for well groups in an extended network that act as a subsea manifold
+    // The wells of such group should have a common THP and total phase rate(s) obeying (if possible) 
+    // the well group constraint set by GCONPROD
+    // subsea manifolds 
+    template <typename TypeTag>
+    void
+    BlackoilWellModel<TypeTag>::
+    computeWellGroupThp(DeferredLogger& local_deferredLogger)
+    {
+        const int reportStepIdx = this->ebosSimulator_.episodeIndex();
+        const auto& network = schedule()[reportStepIdx].network();
+
+        if (!network.active()) {
+            return;
+        }
+
+        auto& well_state = this->wellState();
+        auto& group_state = this->groupState();
 
+        for (const std::string& nodeName : network.node_names()) {
+            const bool has_choke = network.node(nodeName).as_choke();
+            if (has_choke) {
+                const auto& summary_state = this->ebosSimulator_.vanguard().summaryState();
+                const Group& group = schedule().getGroup(nodeName, reportStepIdx);
+                std::optional<Group::ProductionControls> ctrl = group.productionControls(summary_state);
+                const auto cmode = ctrl->cmode;
+                const auto pu = this->phase_usage_;
+                //PJPE: conversion factor res<-> surface rates TODO: to be calculated
+                std::vector<double> resv_coeff(pu.num_phases, 1.0);
+                double gratTargetFromSales = 0.0; //PJPE: check this
+                WellGroupHelpers::TargetCalculator tcalc(cmode, pu, resv_coeff,
+                                                         gratTargetFromSales, nodeName, group_state,
+                                                         group.has_gpmaint_control(cmode));
+                double orig_target = tcalc.groupTarget(ctrl, local_deferredLogger);
+                //PJPE: TODO: include efficiency factor
+                auto mismatch = [&] (auto well_group_thp) {
+                    double well_group_rate(0.0);
+                    for (auto& well : this->well_container_) {
+                        std::string well_name = well->name();
+                        if (group.hasWell(well_name)) {
+                            well->setDynamicThpLimit(well_group_thp);
+                            well->updateWellStateWithTHPTargetProd(this->ebosSimulator_, well_state, local_deferredLogger);
+                            auto& ws = well_state.well(well_name);
+                            well_group_rate += -tcalc.calcModeRateFromRates(ws.surface_rates);
+                        }
+                    }
+                    return well_group_rate - orig_target;
+                };
+
+                // PJPE: TODO what settings to take here?
+                const double tolerance = 1E-4;
```

Comment:
tolerance should probably reflect https://github.com/OPM/opm-common/blob/a5fb8be6835708c8de1b39d76e992e06a526b52b/opm/input/eclipse/Schedule/Network/Balance.hpp#L58C32-L58C32

which is defined as a percentage of the group rate target.